File: frontend/src/server/agents/narrative-builder.ts
import { openai } from "@ai-sdk/openai";
import { generateText } from "ai";
import { env } from "@/env";
import type { AgentInput, AgentOutput } from "@/types/agent";

export class NarrativeBuilderAgent {
  async process(input: AgentInput): Promise<AgentOutput> {
    const startTime = Date.now();

    try {
      if (!env.OPENAI_API_KEY) {
        throw new Error("OPENAI_API_KEY is not configured");
      }

      const topic = input.data.topic as string;
      const facts = input.data.facts as Array<{
        concept: string;
        details: string;
      }>;
      const targetDuration = (input.data.target_duration as number) ?? 60;
      // New optional fields - backward compatible
      const childAge = input.data.child_age as string | null | undefined;
      const childInterest = input.data.child_interest as
        | string
        | null
        | undefined;

      const systemPrompt = this.buildSystemPrompt(
        targetDuration,
        childAge,
        childInterest,
      );
      const userPrompt = this.buildUserPrompt(
        topic,
        facts,
        targetDuration,
        childAge,
        childInterest,
      );

      const result = await generateText({
        model: openai("gpt-4o-mini"),
        system: systemPrompt,
        prompt: userPrompt,
        temperature: 0.7,
      });

      const content = result.text;
      if (!content) {
        throw new Error("No content in LLM response");
      }

      // Extract JSON from response (may be wrapped in markdown code blocks)
      const jsonBlockRegex = /```json\s*([\s\S]*?)\s*```/;
      const jsonObjectRegex = /\{[\s\S]*\}/;
      const jsonBlockMatch = jsonBlockRegex.exec(content);
      const jsonObjectMatch = jsonObjectRegex.exec(content);
      const jsonStr = jsonBlockMatch
        ? (jsonBlockMatch[1] ?? jsonBlockMatch[0])
        : jsonObjectMatch
          ? jsonObjectMatch[0]
          : content;
      const scriptData = JSON.parse(jsonStr) as unknown;

      // Calculate cost (GPT-4o-mini: ~$0.15 per 1M input tokens, ~$0.60 per 1M output tokens)
      const inputTokens = result.usage?.inputTokens ?? 0;
      const outputTokens = result.usage?.outputTokens ?? 0;
      const cost =
        (inputTokens * 0.15) / 1_000_000 + (outputTokens * 0.6) / 1_000_000;

      const duration = (Date.now() - startTime) / 1000;

      return {
        success: true,
        data: { script: scriptData },
        cost,
        duration,
      };
    } catch (error) {
      return {
        success: false,
        data: {},
        cost: 0.0,
        duration: (Date.now() - startTime) / 1000,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  private buildSystemPrompt(
    targetDuration: number,
    childAge?: string | null,
    childInterest?: string | null,
  ): string {
    // Determine age-appropriate language
    let ageContext = "grades 6-7 (reading level ~6.5)";
    let readingLevel = "6.5";

    if (childAge) {
      // Parse age and adjust context
      const ageNum = parseInt(childAge, 10);
      if (!isNaN(ageNum)) {
        if (ageNum < 6) {
          ageContext = "early elementary (ages 4-6, reading level ~2-3)";
          readingLevel = "2.5";
        } else if (ageNum < 9) {
          ageContext = "elementary (ages 6-8, reading level ~3-4)";
          readingLevel = "3.5";
        } else if (ageNum < 12) {
          ageContext = "upper elementary (ages 9-11, reading level ~4-5)";
          readingLevel = "4.5";
        } else if (ageNum < 15) {
          ageContext = "middle school (ages 12-14, reading level ~6-7)";
          readingLevel = "6.5";
        } else {
          ageContext = "high school (ages 15+, reading level ~8-9)";
          readingLevel = "8.5";
        }
      }
    }

    // Build interest context if provided
    const interestContext = childInterest
      ? `\n\nChild's Interest: ${childInterest}\n- Incorporate this interest into examples and connections where relevant\n- Make the content relatable to this interest`
      : "";

    return `You are an expert educational video script creator specializing in creating engaging, accurate educational content for ${ageContext}.

Your task:

1. Create a 4-part educational script (Hook → Concept → Process → Conclusion)
2. Each part has specific timing and purpose
3. Use age-appropriate language for ${ageContext}
4. Include all provided facts and key concepts
5. Provide visual guidance for each segment
6. Ensure scientific accuracy${interestContext}

Structure:

- Hook (0-10s): Engage with question or surprising fact
- Concept Introduction (10-25s): Introduce key vocabulary and ideas
- Process Explanation (25-45s): Explain how/why it works with details
- Conclusion (45-60s): Real-world connection and summary

Rules:

- Total duration must be ${targetDuration} seconds
- Use conversational, enthusiastic tone
- Define technical terms when first introduced
- Include concrete examples
- End with memorable takeaway
- Output ONLY valid JSON, no additional text

Required JSON structure:
{
"total_duration": ${targetDuration},
"reading_level": "${readingLevel}",
"key_terms_count": <number>,
"segments": [
{
"id": "seg_001",
"type": "hook",
"start_time": 0,
"duration": 10,
"narration": "<script text>",
"visual_guidance": "<description of what should be shown>",
"key_concepts": ["<concept1>", "<concept2>"],
"educational_purpose": "<why this segment matters>"
},
{
"id": "seg_002",
"type": "concept_introduction",
"start_time": 10,
"duration": 15,
"narration": "<script text>",
"visual_guidance": "<description>",
"key_concepts": [],
"educational_purpose": "<purpose>"
},
{
"id": "seg_003",
"type": "process_explanation",
"start_time": 25,
"duration": 20,
"narration": "<script text>",
"visual_guidance": "<description>",
"key_concepts": [],
"educational_purpose": "<purpose>"
},
{
"id": "seg_004",
"type": "conclusion",
"start_time": 45,
"duration": 15,
"narration": "<script text>",
"visual_guidance": "<description>",
"key_concepts": [],
"educational_purpose": "<purpose>"
}
]
}`;
  }

  private buildUserPrompt(
    topic: string,
    facts: Array<{ concept: string; details: string }>,
    targetDuration: number,
    childAge?: string | null,
    childInterest?: string | null,
  ): string {
    const factsStr = facts
      .map((f) => `- ${f.concept}: ${f.details}`)
      .join("\n");

    let ageInfo = "";
    if (childAge) {
      ageInfo = `\nChild's Age: ${childAge} years old`;
    }

    let interestInfo = "";
    if (childInterest) {
      interestInfo = `\nChild's Interest: ${childInterest}`;
    }

    return `Topic: ${topic}${ageInfo}${interestInfo}

Key Facts to Include:
${factsStr}

Target Duration: ${targetDuration} seconds

Generate an engaging, scientifically accurate educational script in JSON format.`;
  }
}

File: backend/app/agents/narrative_builder.py
"""
Narrative Builder Agent

This agent generates a structured narrative script for a 60-second video.
It creates a 4-part structure (hook, concept, process, conclusion) with:
- Narration text for each part
- Clip duration suggestions
- Visual guidance for each scene
- Key concepts to highlight
"""

import time
import json
import logging
import uuid
from typing import Optional, List, Dict, Any
import replicate
from app.agents.base import AgentInput, AgentOutput

logger = logging.getLogger(__name__)


class NarrativeBuilderAgent:
    """
    Generates narrative structure and script for educational/promotional videos.

    Input data expected:
    - user_id: str - The ID of the user requesting the narrative
    - topic: str - The main topic/subject of the video
    - learning_objective: str - What the viewer should learn/understand
    - key_points: List[str] - Array of key points to cover

    Output data provided:
    - session_id: str - Unique session ID for this narrative
    - script: dict - The complete script structure with 4 parts
    - cost: float - Cost of the LLM call
    - duration: float - Time taken to generate
    """

    def __init__(self, replicate_api_key: Optional[str] = None):
        """
        Initialize Narrative Builder Agent.
        
        Args:
            replicate_api_key: Replicate API key (defaults to AWS Secrets Manager, then env var)
        """
        # Try to get API key from parameter, then Secrets Manager, then settings
        if replicate_api_key:
            self.api_key = replicate_api_key
        else:
            # Try Secrets Manager first
            try:
                from app.services.secrets import get_secret
                self.api_key = get_secret("pipeline/replicate-api-key")
                logger.debug("Retrieved REPLICATE_API_KEY from AWS Secrets Manager for NarrativeBuilderAgent")
            except Exception as e:
                logger.debug(f"Could not retrieve REPLICATE_API_KEY from Secrets Manager: {e}, falling back to settings")
                from app.config import get_settings
                settings = get_settings()
                self.api_key = settings.REPLICATE_API_KEY
        
        if not self.api_key:
            logger.warning(
                "REPLICATE_API_KEY not set. Narrative generation will fail. "
                "Add it to AWS Secrets Manager (pipeline/replicate-api-key) or .env file."
            )
            self.client = None
        else:
            self.client = replicate.Client(api_token=self.api_key)
        
        self.model = "meta/meta-llama-3-70b-instruct"
        self.cost_per_call = 0.001  # Llama 3 is nearly free

    async def process(self, input: AgentInput) -> AgentOutput:
        """
        Main processing method that generates the narrative script.

        Args:
            input: AgentInput containing user_id, topic, learning_objective, and key_points

        Returns:
            AgentOutput with success status, script data, cost, and duration
        """
        try:
            start_time = time.time()

            # Extract input data
            user_id = input.data.get("user_id")
            topic = input.data.get("topic", "")
            learning_objective = input.data.get("learning_objective", "")
            key_points = input.data.get("key_points", [])

            # Validate inputs
            if not topic:
                raise ValueError("Topic is required")
            if not learning_objective:
                raise ValueError("Learning objective is required")
            if not key_points or len(key_points) == 0:
                raise ValueError("At least one key point is required")

            # Generate a new session ID for this narrative
            session_id = str(uuid.uuid4())

            logger.info(f"[{session_id}] Building narrative for topic: {topic}")
            logger.info(f"[{session_id}] User ID: {user_id}, Key Points: {len(key_points)}")

            # Build prompts
            system_prompt = self._build_system_prompt()
            user_prompt = self._build_user_prompt(topic, learning_objective, key_points)

            # Call LLM to generate narrative
            llm_response = await self._call_llm(system_prompt, user_prompt)

            # Parse the response into structured script
            script = self._parse_script_response(llm_response)

            # Validate script structure
            self._validate_script(script)

            duration = time.time() - start_time

            logger.info(f"[{session_id}] Narrative built successfully in {duration:.2f}s")

            return AgentOutput(
                success=True,
                data={
                    "session_id": session_id,
                    "script": script,
                    "user_id": user_id,
                    "topic": topic,
                    "learning_objective": learning_objective
                },
                cost=self.cost_per_call,
                duration=duration,
                error=None
            )

        except Exception as e:
            logger.error(f"Narrative building failed: {e}", exc_info=True)
            return AgentOutput(
                success=False,
                data={},
                cost=0.0,
                duration=time.time() - start_time,
                error=str(e)
            )

    def _build_system_prompt(self) -> str:
        """Build the system prompt for the LLM."""
        return """You are an expert video script writer and educational content creator. Your task is to create engaging, concise scripts for 60-second educational videos.

You must create a script with EXACTLY 4 parts:
1. HOOK (10-15 seconds) - Grab attention, pose a question or intriguing statement
2. CONCEPT (15-20 seconds) - Introduce the main idea/learning objective
3. PROCESS (20-25 seconds) - Explain how it works, key steps, or main points
4. CONCLUSION (10 seconds) - Summarize key takeaway and call-to-action

For each part, you must provide:
- narration: The exact script text to be spoken
- duration: Suggested duration in seconds
- visual_guidance: Description of what should be shown on screen
- key_concepts: Array of 1-3 key terms/concepts to highlight visually

The narration should be:
- Conversational and engaging
- Clear and concise
- Suitable for voice-over
- Timed appropriately for the duration

The visual guidance should be:
- Specific and actionable
- Describe scenes, graphics, text overlays, or animations
- Support the narration

You MUST respond with valid JSON in this exact structure:
{
  "hook": {
    "narration": "string",
    "duration": number,
    "visual_guidance": "string",
    "key_concepts": ["string"]
  },
  "concept": {
    "narration": "string",
    "duration": number,
    "visual_guidance": "string",
    "key_concepts": ["string"]
  },
  "process": {
    "narration": "string",
    "duration": number,
    "visual_guidance": "string",
    "key_concepts": ["string"]
  },
  "conclusion": {
    "narration": "string",
    "duration": number,
    "visual_guidance": "string",
    "key_concepts": ["string"]
  }
}

Do not include any text before or after the JSON object."""

    def _build_user_prompt(
        self,
        topic: str,
        learning_objective: str,
        key_points: List[str]
    ) -> str:
        """Build the user prompt with specific topic and requirements."""
        key_points_formatted = "\n".join([f"- {point}" for point in key_points])

        return f"""Create a 60-second video script about the following topic:

TOPIC: {topic}

LEARNING OBJECTIVE: {learning_objective}

KEY POINTS TO COVER:
{key_points_formatted}

Remember to create an engaging hook, clearly explain the concept, walk through the process/key points, and end with a strong conclusion. The total duration should be approximately 60 seconds across all 4 parts.

Respond with ONLY the JSON object, no additional text."""

    async def _call_llm(self, system_prompt: str, user_prompt: str) -> str:
        """
        Call the Replicate LLM API.

        Args:
            system_prompt: System instructions for the LLM
            user_prompt: User query/request

        Returns:
            str: The LLM's response
        """
        if not self.client:
            raise ValueError("REPLICATE_API_KEY not configured. Set it in AWS Secrets Manager (pipeline/replicate-api-key) or .env file.")
        
        logger.info("Calling Replicate API with Llama 3 70B...")

        output = await self.client.async_run(
            self.model,
            input={
                "system_prompt": system_prompt,
                "prompt": user_prompt,
                "max_tokens": 2000,
                "temperature": 0.7,
                "top_p": 0.9,
                "top_k": 50
            }
        )

        # Concatenate streaming output
        full_response = "".join([chunk for chunk in output])

        logger.info(f"Received LLM response: {len(full_response)} characters")

        return full_response

    def _parse_script_response(self, response: str) -> dict:
        """
        Parse the LLM response to extract JSON script structure.

        Args:
            response: Raw LLM response string

        Returns:
            dict: Parsed script structure
        """
        # Clean up response
        response = response.strip()

        # Find JSON object boundaries
        start_idx = response.find("{")
        end_idx = response.rfind("}") + 1

        if start_idx == -1 or end_idx == 0:
            raise ValueError("No JSON object found in LLM response")

        json_str = response[start_idx:end_idx]

        try:
            script = json.loads(json_str)
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse JSON: {e}")
            logger.error(f"JSON string: {json_str[:500]}...")
            raise ValueError(f"Invalid JSON in LLM response: {e}")

        return script

    def _validate_script(self, script: dict) -> None:
        """
        Validate that the script has all required parts and fields.

        Args:
            script: The parsed script dictionary

        Raises:
            ValueError: If script is missing required parts or fields
        """
        required_parts = ["hook", "concept", "process", "conclusion"]
        required_fields = ["narration", "duration", "visual_guidance", "key_concepts"]

        # Check all parts exist
        for part in required_parts:
            if part not in script:
                raise ValueError(f"Script missing required part: {part}")

            # Check all fields exist in each part
            for field in required_fields:
                if field not in script[part]:
                    raise ValueError(f"Script part '{part}' missing required field: {field}")

            # Validate types
            if not isinstance(script[part]["narration"], str):
                raise ValueError(f"Narration in '{part}' must be a string")

            if not isinstance(script[part]["duration"], (int, float)):
                raise ValueError(f"Duration in '{part}' must be a number")

            if not isinstance(script[part]["visual_guidance"], str):
                raise ValueError(f"Visual guidance in '{part}' must be a string")

            if not isinstance(script[part]["key_concepts"], list):
                raise ValueError(f"Key concepts in '{part}' must be an array")

        # Check total duration is reasonable (45-75 seconds acceptable range)
        total_duration = sum(
            script[part]["duration"] for part in required_parts
        )

        if total_duration < 45 or total_duration > 75:
            logger.warning(
                f"Total script duration {total_duration}s is outside recommended range (45-75s)"
            )
